### PR TITLE
mobile direct binding support for wallet unlocker

### PR DESCRIPTION
**Pull request by biscottigelato**
I am a Go newbie. Hence why i figure I should start with this relatively small task (even tho still took me a while @@) Hopefully this isn't a waste of your time. There is way more copy/paste/code duplication than I'd like. But short of that it seem quite difficult since LightningClient and WalletUnlockerClient are ultimately distinct interfaces? If there is any other way that you can think of, just gimme a keyword and I'll try to figure out how to do so on my own.

The new interfaces are tested to work at least in my iOS app environment. I also removed the 3 explicitly coded commands in bindings.go. Not sure why you weren't just letting those be automatically generated via gen_bindings?

### CODE REVIEW

File: daemon/lnd.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ var (
 // lndMain is the true entry point for lnd. This function is required since
 // defers created in the top-level scope of a main method aren't executed if
 // os.Exit() is called.
-func LndMain(appDir string, lis net.Listener) error {
+func LndMain(appDir string, lightningLis net.Listener, unlockerLis net.Listener) error {
```
Comment (halseth) [inline daemon/lnd.go:94]: think it should be possible to use the same listener for both..?

File: mobile/gen_bindings.go [hunk 2]
```diff
@@ -162,7 +151,20 @@ func main() {
...
 			switch {
+			case s.GetName() == "WalletUnlocker":
+
+				p := rpcParams{
+					MethodName:  m.GetName(),
+					RequestType: m.GetInputType()[1:],
+					Comment:     godoc[name],
+				}
+
+				if err := unlockOnceTemplate.Execute(wr, p); err != nil {
```
Comment (halseth) [inline mobile/gen_bindings.go:162]: We ideally should reuse the same template, just passing in the service name

File: mobile/memrpc.go [hunk 3]
```diff
@@ -98,13 +105,95 @@ func getClient() (lnrpc.LightningClient, context.Context, func(), error) {
...
+
+	// getSync calls the desired method on the given client in a
+	// blocking matter.
+	getSync func(context.Context, lnrpc.WalletUnlockerClient,
+		proto.Message) (proto.Message, error)
+}
+
+// start executes the RPC call specified by this onceHandler using the
+// specified serialized msg request.
+func (s *unlockOnceHandler) start(msg []byte, callback Callback) {
```
Comment (halseth) [inline mobile/memrpc.go:152]: same, should try to reuse as much as possible :)

**Review by halseth — CHANGES_REQUESTED** [review]
Thanks! 😀

File: mobile/memrpc.go (hunk 3 above)
Comment (biscottigelato) [reply mobile/memrpc.go:152]: Hey Thanks for looking at this! My struggle is with mixing the 2 return interface types together LightningClient & WalletUnlockerClient. Any go construct I should look at for this? Still a newbie at go for sure. I'll regardless revisit this in the next few days to see if I can come up with anything better to reduce redundancies. But any tips would be appreciated!

**Comment by biscottigelato** [comment]
Just want to update that I do plan to work on this some more. But I am trying to rebase this to the tip of lightningnetwork/lnd @ master before proceeding any further. I got the rebase done but hitting a problem where the tip of lnd is crashing any btcd node it touches. So need to wait for that to resolve before proceeding any further on this.

https://github.com/btcsuite/btcd/issues/1219

**Comment by caffeinum** [comment]
@halseth @biscottigelato what about updating this to lnd v5 ?

**Comment by halseth** [comment]
This is now in master

**Pull request closed without merging**